### PR TITLE
Implementation of a more flexible optimizer/scheduler override system

### PR DESCRIPTION
**Pull request by jstjohn**
# What does this PR do ?
This PR removes the hard-coded wd_mult, and brings back the flexibility of the previous lambda API while allowing users to override a number of different optimizer or scheduler configs per layer. This ideas were discussed in the design document "Optimizer param lookup refactor idea".

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -1171,15 +1174,19 @@ def get_megatron_optimizer_config(args: Any) -> OptimizerConfig:
...
-    else:
-        config_overrides = None
+            decoupled_lr_config["min_lr"] = args.decoupled_min_lr
+        config_overrides[decoupled_param_key] = decoupled_lr_config
+    
+    # Next construct the standard param group overrides for no weight decay on bias parameters 
+    #  as well as any length 1 parameters.
+    param_length_1_match = ParamPredicate(name="param_len_1", fn=lambda param: len(param.shape) == 1)
+    param_wd_mult_key = ParamKey(name="*.bias", predicate=param_length_1_match)
+    config_overrides[param_wd_mult_key] = ParamGroupOverride(wd_mult=0.0)
```
Comment (jstjohn) [inline megatron/training/training.py]: @ko3n1g this is the code block where I am implementing the previous weight decay rule. Hopefully this is flexible enough for the model you are blocked on. LMK if not.

File: tests/unit_tests/test_optimizer.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 
 import os
+from ast import Not
```
Comment (jstjohn) [inline tests/unit_tests/test_optimizer.py]: This was not intentional. I'll remove it.

File: tests/unit_tests/test_utilities.py [hunk 3]
```diff
@@ -27,8 +27,8 @@ def __init__(
 
 class Utils:
 
-    world_size = int(os.environ['WORLD_SIZE'])
-    rank = int(os.environ['LOCAL_RANK'])
+    world_size = int(os.environ.get('WORLD_SIZE', '1'))
+    rank = int(os.environ.get('LOCAL_RANK', '0'))
```
Comment (jstjohn) [inline tests/unit_tests/test_utilities.py:32]: These changes allow me to directly run `pytest tests/unit_tests/test_optimizer_config.py` for example which do not actually depend on being run in a torchrun/slurm env.

File: megatron/training/training.py (hunk 1 above)
Comment (deepakn94) [reply megatron/training/training.py]: I wonder if this should go into MCore somehow to make sure all consumers of MCore don't see changed default behavior?

**Review by deepakn94 — COMMENTED** [review]
This looks great. Thanks a lot, @jstjohn!

Just one minor comment.

File: megatron/training/training.py (hunk 1 above)
Comment (jstjohn) [reply megatron/training/training.py]: Done. So now when a user does not supply any overrides (which does not happen in the training script, but would happen in a legacy training loop) we do the standard weight decay skip. For handling decoupled LR, we already change the API so those users will see that they need to change their loops to make use of the new API, and hopefully will use the megatron trainer as a reference.

**Comment by ko3n1g** [comment]
/ok to test [6ae2c41](https://github.com/NVIDIA/Megatron-LM/pull/2723/commits/6ae2c4105d3c7902874075d420ff76a015d28ba1)

**Comment by ko3n1g** [comment]
/ok to test [8ca2cf4](https://github.com/NVIDIA/Megatron-LM/pull/2723/commits/8ca2cf4acd964e435078bc8343bb02c5439fd3d9)

**Review by yaoyu-33 — APPROVED** [review]
